keps/168-2-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,313 @@
+
+### API endpoints:
+
+We introduce a new API group ```visibility.kueue.x-k8s.io``` that aggregates following endpoints: 
+
+#### List all pending workloads in ClusterQueue
+
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME/pending_workloads?offset=0&limit=1000
+```
+
+#### List all pending workloads in LocalQueue
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/namespaces/LQ_NAMESPACE/localqueues/LQ_NAME/pending_workloads?offset=0&limit=1000
+```
+
+#### Fetch information about a single pending Workload
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/namespaces/WL_NAMESPACE/pending_workload/WL_NAME
+```
+
+### API Objects:
+
+```
+// PendingWorkloadSummary is a user-facing representation of a pending workload in both LocalQueues and ClusterQueue that summarizes neccessary information from the admission order perspective
+type PendingWorkloadSummary struct {
+	TypeMeta TypeMeta
+	ObjectMeta ObjectMeta
+
+	LocalQueueName string
+	PositionInClusterQueue int32
+	PositionInLocalQueue int32
+	Priority int32
+}
+
+// PendingWorkloadSummaryList contains a list of pending workloads in the context
+// of the query (within LocalQueue or ClusterQueue).
+type PendingWorkloadsSummaryList struct {
+	TypeMeta TypeMeta
+	ListMeta ListMeta
+
+	Items []PendingWorkloadSummaryInClusterQueue
+}
+```
+
+### Future extensions
+
+The introduced API uses mechanism of subresources. It means, that in the future it can be easily extended by adding additional endpoints related e.g. to admitted workloads. Potentially the endpoint could look like this:
+
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME/admitted_workloads?offset=0&limit=1000
+```
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Overview
+
+Our main focus is integration tests, as most of the added code is responsible for integrating with the Kueue and RBAC roles.
+
+#### Unit Tests
+
+We plan on adding unit tests that cover getting a list of pending workloads at the KueueManager level.
+
+- `pkg/visibility`: `30 Oct 2023` - `0%`
+
+#### Integration tests
+
+Integration tests should check if our server work correctly according to the assumptions we mentioned:
+- Pending workloads are returned according to their actual status without delay.
+- Adding workloads to one LocalQueues results in position changes for workloads submitted to other LocalQueues,
+- Data is consistent across all LocalQueues
+- User with only LocalQueue visibility cannot access the list of pending workloads for ClusterQueue
+
+#### E2E tests
+
+We plan on adding sanity e2e tests.
```

Comment:
Add that we also want to use e2e tests for RBAC. You may list some of the critical scenarios:
- clusters queues can only be accessed by admin users
- local queues can be accessed by only users with the visibility to the corresponding namespaces